firefox_ui_harness/runtests.py

```diff
@@ -7,71 +7,57 @@
 
 
-def startTestRunner(runner_class, options, tests):
-    install_folder = None
-
-    try:
-        # Prepare the workspace path so that all temporary data can be stored inside it.
-        if options.workspace_path:
-            path = os.path.expanduser(options.workspace_path)
-            options.workspace = os.path.abspath(path)
-
-            if not os.path.exists(options.workspace):
-                os.makedirs(options.workspace)
-        else:
-            options.workspace = tempfile.mkdtemp('.{}'.format(os.path.basename(sys.argv[0])))
-
-        options.logger.info('Using workspace for temporary data: "{}"'.format(options.workspace))
+class FirefoxUIHarness(MarionetteHarness):
+    def __init__(self,
+                 runner_class=FirefoxUITestRunner,
+                 parser_class=FirefoxUIArguments):
+        MarionetteHarness.__init__(self, runner_class, parser_class)
+        self.install_folder = None
 
+    def process_args(self):
+        # TODO move installer to mozharness script.
         # If the specified binary is an installer it needs to be installed
-        if options.installer:
-            installer = os.path.realpath(options.installer)
-
-            dest_folder = os.path.join(options.workspace, 'binary')
-            options.logger.info('Installing application "%s" to "%s"' % (installer,
-                                                                         dest_folder))
-            install_folder = mozinstall.install(installer, dest_folder)
-            options.binary = mozinstall.get_binary(install_folder, 'firefox')
-
-        runner = runner_class(**vars(options))
-        runner.run_tests(tests)
-
-    finally:
-        # Ensure to uninstall the binary if it has been installed before
-        if install_folder and os.path.exists(install_folder):
-            options.logger.info('Uninstalling application at "%s"' % install_folder)
-            mozinstall.uninstall(install_folder)
-
-    return runner
-
-
-def cli(runner_class=FirefoxUITestRunner, parser_class=FirefoxUIArguments):
-    parser = parser_class(usage='%(prog)s [options] test_file_or_dir <test_file_or_dir> ...')
-    structured.commandline.add_logging_group(parser)
-    args = parser.parse_args()
-    parser.verify_usage(args)
-
-    logger = structured.commandline.setup_logging(
-        args.logger_name, args, {'mach': sys.stdout})
-    args.logger = logger
-
-    try:
-        runner = startTestRunner(runner_class, args, args.tests)
-        if runner.failed > 0:
-            sys.exit(10)
-
-    except Exception:
-        logger.error('Failure during execution of the update test.',
-                     exc_info=True)
-        sys.exit(1)
-
+        if self.args.installer:
+            installer = os.path.realpath(self.args.installer)
+
+            if not self.args.workspace:
+                self.args.installer_workspace = tempfile.mkdtemp(
+                    '.{}'.format(os.path.basename(sys.argv[0]))
```

Comment:
A giant summary:
* `--workspace`: nothing in the workspace dir is ever deleted by Marionette runner, except profile clones (see --profile below), because the idea is that you might want to examine the content of the workspace to investigate test failures. If you specify a workspace, temporary profiles will be stored there for later inspection. It's the consumer's responsibility to clear the workspace.  
* `--installer`: in the absence of a workspace, if we create a temp dir to hold the installed binaries, then it's still our responsibility to delete that temp dir. If we "totally forget about a workspace by default" for the binaries, they will just go in cwd. I think it's better to create a temp dir and then clean it up.
* `--profile`: this is used to specify the __source__ of a profile to copy, say `./awesome_profile`. Copies of `awesome_profile` are made in TMP or the workspace (if specified), and then they are deleted, since you will still have `./awesome_profile` available if you want to examine it. There's a special case: when `./awesome_profile` is empty to begin with. Then the "clones" are filed with fresh, temporary profile data, but they still get deleted in the end. In other words, if you want to use `--workspace` but you don't want to keep the profiles, you can use `--profile` to specify an empty or non-existant directory.

So, for example, if you run `firefox-ui-tests --installer blah --workspace build --profile build/some_empty_dir` at the end you will have a directory called `build` that contains `binary` (empty), `some_empty_dir` (empty), `gecko.log` and nothing else because the clones of `some_empty_dir` all get deleted from the workspace. 

Your decisions:
* How is the workspace treated in mozmill-ci for firefox-ui-tests? Do you delete it after each job? Do you keep it around for investigations?
* Do you want to keep profiles for investigation or always delete them immediately?